staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go

```diff
@@ -141,22 +167,43 @@ func MergeAPIResourceConfigs(
 		if err != nil {
 			return nil, err
 		}
-		if enabled {
+
+		switch len(tokens) {
+		case 2:
+			versionPreferences = append(versionPreferences, versionEnablementPreference{
+				key:          key,
+				enabled:      enabled,
+				groupVersion: groupVersion,
+			})
+		case 3:
+			resourcePreferences = append(resourcePreferences, resourceEnablementPreference{
+				key:                  key,
+				enabled:              enabled,
+				groupVersionResource: groupVersion.WithResource(tokens[2]),
+			})
+		}
+	}
+
+	// apply version preferences first, so that we can remove the hardcoded resource preferences that are being overridden
+	for _, versionPreference := range versionPreferences {
+		// if a user has expressed a preference about a version, that preference takes priority over the hardcoded resources
+		resourceConfig.RemoveMatchingResourcePreferences(resourceMatcherForVersion(versionPreference.groupVersion))
+		if versionPreference.enabled {
 			// enable the groupVersion for "group/version=true" and "group/version/resource=true"
-			resourceConfig.EnableVersions(groupVersion)
-		} else if len(tokens) == 2 {
+			resourceConfig.EnableVersions(versionPreference.groupVersion)
+
+		} else {
 			// disable the groupVersion only for "group/version=false", not "group/version/resource=false"
-			resourceConfig.DisableVersions(groupVersion)
+			resourceConfig.DisableVersions(versionPreference.groupVersion)
 		}
+	}
 
-		if len(tokens) < 3 {
-			continue
-		}
-		groupVersionResource := groupVersion.WithResource(tokens[2])
-		if enabled {
-			resourceConfig.EnableResources(groupVersionResource)
+	// apply resource preferences last, so they have the highest priority
+	for _, resourcePreference := range resourcePreferences {
+		if resourcePreference.enabled {
+			resourceConfig.EnableResources(resourcePreference.groupVersionResource)
```

Comment:
looking at ResourceConfig#ResourceEnabled, I don't think this is effective if the group/version is not enabled

I think that has to change to
```diff
 func (o *ResourceConfig) ResourceEnabled(resource schema.GroupVersionResource) bool {
-       if !o.VersionEnabled(resource.GroupVersion()) {
-               return false
-       }
        resourceEnabled, explicitlySet := o.ResourceConfigs[resource]
-       if !explicitlySet {
-               return true
+       if explicitlySet {
+               return resourceEnabled
        }
-       return resourceEnabled
+       return o.VersionEnabled(resource.GroupVersion())
 }
```

If this wasn't actually effective, needs test coverage for when the group/version is explicitly or implicitly off, and a specific resource is explicitly on

Reply:
mm... I think AnyVersionForGroupEnabled is also incorrect in the presence of an otherwise-disabled version but enabled resource within that version

Reply:
Updated.